### PR TITLE
strongswan: Add dependency to virtual package strongswan-mod-socket

### PR DESCRIPTION
**Pull request by pprindeville**
Maintainer: me, @Thermi 
Compile tested: x86_64, generic, HEAD (f038a169a0)
Run tested: same, installed on test VM

Description:

Originally by Noel Kuntze, and refined by Stijn Tintel.

### CODE REVIEW

**Comment by stintel** [comment]
You need to pull in my latest changes. I was still building an image for runtime test, and was about to open another PR then.

File: net/strongswan/Makefile [hunk 1]
```diff
@@ -441,10 +441,17 @@ $(call Package/strongswan/description/Default)
...
 define BuildPlugin
+  $(call BuildPluginProvides $1,$2,$3,)
+endef
+
+define BuildPluginProvides
   define Package/strongswan-mod-$(1)
     $$(call Package/strongswan/Default)
     TITLE:= StrongSwan $(2) plugin
     DEPENDS:= strongswan $(3)
+ifneq ($(4),)
```
Comment (pprindeville) [inline net/strongswan/Makefile:452]: Can we just error out if this version is used but no 4th arg is provided?

**Comment by CodeFetch** [comment]
Now I have an issue here. mod-socket and mod-socket-dynamic are conflicting. When selecting both I needed to delete the default library to use the dynamic one. Thus mod-socket-dynamic needs to be created as an individual package and PROVIDE mod-socket...

File: net/strongswan/Makefile (hunk 1 above)
Comment (pprindeville) [reply net/strongswan/Makefile:452]: i.e.

```
ifeq ($(4),)
    $$(error BuildPluginProvides requires provider suffix)
endif
```

Comment (pprindeville) [reply net/strongswan/Makefile:448]: Amended

**Comment by pprindeville** [comment]
@stintel @Thermi Can you please test this PR?

**Comment by pprindeville** [comment]
@CodeFetch @Neustradamus Can someone please test this?

**Comment by Neustradamus** [comment]
There is a progress on this PR?

**Comment by pprindeville** [comment]
> There is a progress on this PR?

Still need it tested.

**Comment by Thermi** [comment]
I'd say it works

**Comment by stintel** [comment]
Why was this merged, even though I said in #16373 it was no good? It causes the following problem:

```
tmp/.config-package.in:122354:error: recursive dependency detected!
tmp/.config-package.in:122354:  symbol PACKAGE_strongswan-default depends on PACKAGE_strongswan-mod-socket-default
tmp/.config-package.in:123534:  symbol PACKAGE_strongswan-mod-socket-default is selected by PACKAGE_strongswan-default
```

**Comment by Thermi** [comment]
Thank you @stintel!

I guess we were too quick with the merge. I took some time this morning to take a look and the issue disappears if we just use the virtual provider instead of socket-default. The error message is not useful tbh  because it does not describe a circular dependency.

I still need to check if doing that change doesn't cause unsatisfiable dependencies.

Am 16. August 2022 12:21:59 UTC schrieb Stijn Tintel ***@***.***>:
>Why was this merged, even though I said in #16373 it was no good? It causes the following problem:
>
>```
>tmp/.config-package.in:122354:error: recursive dependency detected!
>tmp/.config-package.in:122354:  symbol PACKAGE_strongswan-default depends on PACKAGE_strongswan-mod-socket-default
>tmp/.config-package.in:123534:  symbol PACKAGE_strongswan-mod-socket-default is selected by PACKAGE_strongswan-default
>```
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/openwrt/packages/pull/16367#issuecomment-1216562245
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Sent from mobile

**Comment by stintel** [comment]
That whole circular dependency stuff is a disaster. It's the main reason I gave up on #16373 as I could just select the proper package in my .config and get on with life.

**Comment by Thermi** [comment]
I fixed the issue, see #19184

**Pull request merged**